File: backend/app/schemas/finance.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel


class FinanceValidationRow(BaseModel):
    row_number: int
    data: dict
    status: str  # valid | error | warning
    errors: list[str]
    warnings: list[str]


class FinanceUploadValidationResponse(BaseModel):
    total_rows: int
    valid_count: int
    error_count: int
    duplicate_count: int
    warning_count: int
    period: str
    version: int
    rows: list[FinanceValidationRow]
    upload_token: str


class FinanceUploadConfirmRequest(BaseModel):
    upload_token: str


class FinanceUploadConfirmResponse(BaseModel):
    imported_count: int
    version: int
    message: str


class FinanceBillableResponse(BaseModel):
    id: str
    employee_id: str
    employee_name: str
    period: str
    billable_status: str
    billable_hours: float
    billed_amount: Optional[float] = None
    project_name: Optional[str] = None
    client_name: Optional[str] = None
    version: int


class FinanceBillableListResponse(BaseModel):
    entries: list[FinanceBillableResponse]
    total: int
    period: str
    latest_version: int

File: frontend/src/types/finance.ts
export interface FinanceBillableEntry {
  id: string
  employee_id: string
  employee_name: string
  period: string
  billable_status: string
  billable_hours: number
  billed_amount?: number
  project_name?: string
  client_name?: string
  version: number
}

export interface FinanceValidationRow {
  row_number: number
  data: Record<string, string>
  status: "valid" | "error" | "warning"
  errors: string[]
  warnings: string[]
}

export interface FinanceUploadValidationResponse {
  total_rows: number
  valid_count: number
  error_count: number
  duplicate_count: number
  warning_count: number
  period: string
  version: number
  rows: FinanceValidationRow[]
  upload_token: string
}

export interface FinanceBillableListResponse {
  data_source?: "hrms" | "excel"
  entries: FinanceBillableEntry[]
  total: number
  period: string
  latest_version: number
}

export interface FinanceUploadHistoryEntry {
  batch_id: string
  period: string
  uploaded_by: string
  filename: string
  total_rows: number
  valid_count: number
  error_count: number
  duplicate_count: number
  version: number
  errors: { row: number; message: string }[]
  uploaded_at: string
}
